File: app/src/main/java/com/vteam/testdemo/chat/ChatActivity.java
package com.vteam.testdemo.chat;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.OnProgressListener;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.StorageTask;
import com.google.firebase.storage.UploadTask;
import com.vteam.testdemo.R;
import com.vteam.testdemo.chat.adapter.MessageAdapter;
import com.vteam.testdemo.common.Constants;
import com.vteam.testdemo.common.Utils;
import com.vteam.testdemo.landing.model.ChatModel;
import com.vteam.testdemo.top.Messages;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import de.hdodenhof.circleimageview.CircleImageView;

import static com.vteam.testdemo.common.Constants.NODES.CHAT_NODE;
import static com.vteam.testdemo.common.Constants.NODES.MESSAGES_NODE;

public class ChatActivity extends AppCompatActivity {
    private final List<Messages> messagesList = new ArrayList<>();
    private String messageReceiverID, messageReceiverName, messageReceiverImage, messageSenderID;
    private TextView userName, userLastSeen;
    private CircleImageView userImage;
    private Toolbar ChatToolBar;
    private FirebaseAuth mAuth;
    private DatabaseReference RootRef;
    private ImageButton SendMessageButton, SendFilesButton;
    private EditText MessageInputText;
    private LinearLayoutManager linearLayoutManager;
    private MessageAdapter messageAdapter;
    private RecyclerView userMessagesList;

    private ProgressDialog loadingBar;


    private String saveCurrentTime, saveCurrentDate;
    private String checker = "", myUrl = "";
    private Uri fileUri;
    private StorageTask uploadTask;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);
        setActionBar();

        mAuth = FirebaseAuth.getInstance();
        messageSenderID = mAuth.getCurrentUser().getUid();
        RootRef = FirebaseDatabase.getInstance().getReference();


        Bundle extras = getIntent().getExtras();
        messageReceiverID = extras.get("visit_user_id").toString();
        messageReceiverName = extras.get("visit_user_name").toString();
        if(extras.containsKey("visit_image") && extras.get("visit_image")!=null) {
            messageReceiverImage = extras.get("visit_image").toString();
        }


        IntializeControllers();


        userName.setText(messageReceiverName);


        StorageReference reference = FirebaseStorage.getInstance().getReference();

        if (!TextUtils.isEmpty(messageReceiverImage)) {
            reference.child(messageReceiverImage).getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                @Override
                public void onSuccess(Uri uri) {
                    // Got the download URL for 'users/me/profile.png'
                    Log.d("URL", "" + uri);
                    Glide.with(getApplicationContext())
                            .load(uri)
                            .into(userImage);
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception exception) {
                    // Handle any errors
                }
            });
        }

        SendMessageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SendMessage();
            }
        });
        DisplayLastSeen();
        SendFilesButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                CharSequence options[] = new CharSequence[]
                        {
                                "Images",
                                "Pdf Files",
                                "MS Word Files"
                        };
                AlertDialog.Builder builder = new AlertDialog.Builder(ChatActivity.this);
                builder.setTitle("Select the File");
                builder.setItems(options, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        if (i == 0) {
                            checker = "image";
                            Intent intent = new Intent();
                            intent.setAction(Intent.ACTION_GET_CONTENT);
                            intent.setType("image/*");
                            startActivityForResult(intent.createChooser(intent, "Select Image"), 443);
                        }
                        if (i == 1) {
                            checker = "pdf";
                            Intent intent = new Intent();
                            intent.setAction(Intent.ACTION_GET_CONTENT);
                            intent.setType("application/pdf");
                            startActivityForResult(intent.createChooser(intent, "Select PDF"), 443);
                        }
                        if (i == 2) {
                            checker = "docx";
                            Intent intent = new Intent();
                            intent.setAction(Intent.ACTION_GET_CONTENT);
                            intent.setType("application/msword");
                            startActivityForResult(intent.createChooser(intent, "Select MSWORD FILE"), 443);
                        }
                    }
                });
                builder.show();
            }
        });
    }

    private void IntializeControllers() {


        loadingBar = new ProgressDialog(this);

        userName = findViewById(R.id.custom_profile_name);
        userImage = findViewById(R.id.custom_profile_image);
        userLastSeen = findViewById(R.id.custom_user_last_seen);

        SendMessageButton = findViewById(R.id.send_message_btn);
        SendFilesButton = findViewById(R.id.send_files_btn);
        MessageInputText = (EditText) findViewById(R.id.input_message);

        messageAdapter = new MessageAdapter(messagesList);
        userMessagesList = (RecyclerView) findViewById(R.id.private_messages_list_of_users);
        linearLayoutManager = new LinearLayoutManager(this);
        userMessagesList.setLayoutManager(linearLayoutManager);
        userMessagesList.setAdapter(messageAdapter);

        Calendar calendar = Calendar.getInstance();

        SimpleDateFormat currentDate = new SimpleDateFormat("MMM dd, yyyy");
        saveCurrentDate = currentDate.format(calendar.getTime());

        SimpleDateFormat currentTime = new SimpleDateFormat("hh:mm a");
        saveCurrentTime = currentTime.format(calendar.getTime());

    }

    private void setActionBar() {
        ChatToolBar = (Toolbar) findViewById(R.id.chat_toolbar);
        setSupportActionBar(ChatToolBar);

        ActionBar actionBar = getSupportActionBar();
//        actionBar.setDefaultDisplayHomeAsUpEnabled(true);
        actionBar.setDisplayShowCustomEnabled(true);

        LayoutInflater layoutInflater = (LayoutInflater) this.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View actionBarView = layoutInflater.inflate(R.layout.custom_chat_bar, null);
        actionBar.setCustomView(actionBarView);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 443 && resultCode == RESULT_OK && data != null && data.getData() != null) {
            loadingBar.setTitle("Sending File");
            loadingBar.setMessage("Please wait, we are sending....");
            loadingBar.setCanceledOnTouchOutside(false);
            loadingBar.show();

            fileUri = data.getData();
            if (!checker.equals("image")) {
                StorageReference storageReference = FirebaseStorage.getInstance().getReference().child("Image Files");
                final String messageSenderRef = "Messages/" + messageSenderID + "/" + messageReceiverID;
                final String messageReceiverRef = "Messages/" + messageReceiverID + "/" + messageSenderID;

                DatabaseReference userMessageKeyRef = RootRef.child("Messages")
                        .child(messageSenderID).child(messageReceiverID).push();

                final String messagePushID = userMessageKeyRef.getKey();

                final StorageReference filePath = storageReference.child(messagePushID + "." + checker);

                filePath.putFile(fileUri).addOnCompleteListener(new OnCompleteListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<UploadTask.TaskSnapshot> task) {
                        if (task.isSuccessful()) {
                            Map messageTextBody = new HashMap();
                            messageTextBody.put("message", myUrl);
                            messageTextBody.put("name", fileUri.getLastPathSegment());
                            if (checker.equals("pdf")) {
                                messageTextBody.put("type", checker);
                            } else {
                                messageTextBody.put("type", checker);
                            }

                            messageTextBody.put("from", messageSenderID);
                            messageTextBody.put("to", messageReceiverID);
                            messageTextBody.put("messageID", messagePushID);
                            messageTextBody.put("time", saveCurrentTime);
                            messageTextBody.put("date", saveCurrentDate);

                            Map messageBodyDetails = new HashMap();
                            messageBodyDetails.put(messageSenderRef + "/" + messagePushID, messageTextBody);
                            messageBodyDetails.put(messageReceiverRef + "/" + messagePushID, messageTextBody);

                            RootRef.updateChildren(messageBodyDetails);
                            loadingBar.dismiss();
                        }
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        loadingBar.dismiss();
                        Toast.makeText(ChatActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                }).addOnProgressListener(new OnProgressListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onProgress(UploadTask.TaskSnapshot taskSnapshot) {
                        double p = (100.0 * taskSnapshot.getBytesTransferred()) / taskSnapshot.getTotalByteCount();
                        loadingBar.setMessage((int) p + "% Uploading...");

                    }
                });
            } else if (checker.equals("image")) {
                StorageReference storageReference = FirebaseStorage.getInstance().getReference().child("Image Files");
                final String messageSenderRef = "Messages/" + messageSenderID + "/" + messageReceiverID;
                final String messageReceiverRef = "Messages/" + messageReceiverID + "/" + messageSenderID;

                DatabaseReference userMessageKeyRef = RootRef.child("Messages")
                        .child(messageSenderID).child(messageReceiverID).push();

                final String messagePushID = userMessageKeyRef.getKey();

                final StorageReference filePath = storageReference.child(messagePushID + "." + "jpg");

                uploadTask = filePath.putFile(fileUri);
                uploadTask.continueWithTask(new Continuation() {
                    @Override
                    public Object then(@NonNull Task task) throws Exception {
                        if (!task.isSuccessful()) {
                            throw task.getException();
                        }
                        return filePath.getDownloadUrl();
                    }
                }).addOnCompleteListener(new OnCompleteListener<Uri>() {
                    @Override
                    public void onComplete(@NonNull Task<Uri> task) {
                        Uri downloadUrl = task.getResult();
                        myUrl = downloadUrl.toString();
                        String messageSenderRef = "ChatNode/" + messageSenderID + "/" + messageReceiverID;
                        String messageReceiverRef = "ChatNode/" + messageReceiverID + "/" + messageSenderID;
                        String messageRef = "MessagesNode/" + Utils.setOneToOneChat(messageSenderID,messageReceiverID);


                        ChatModel chatModel = new ChatModel();
                        chatModel.setFromUid(messageSenderID);
                        chatModel.setFromUid(messageReceiverID);
                        chatModel.setDate(saveCurrentDate);
                        chatModel.setTime(saveCurrentTime);
                        chatModel.setLastMessage(myUrl);
                        chatModel.setType(checker);

                        Map messageTextBody = new HashMap();
                        messageTextBody.put("message", myUrl);
                        messageTextBody.put("name", fileUri.getLastPathSegment());
                        messageTextBody.put("type", checker);
                        messageTextBody.put("from", messageSenderID);
                        messageTextBody.put("to", messageReceiverID);
                        messageTextBody.put("messageID", messagePushID);
                        messageTextBody.put("time", saveCurrentTime);
                        messageTextBody.put("date", saveCurrentDate);

                        Map messageBodyDetails = new HashMap();
                        messageBodyDetails.put(messageSenderRef + "/" , chatModel);
                        messageBodyDetails.put(messageReceiverRef + "/" + messagePushID, chatModel);
                        messageBodyDetails.put(messageRef + "/" + messagePushID, messageTextBody);


                        RootRef.updateChildren(messageBodyDetails).addOnCompleteListener(new OnCompleteListener() {
                            @Override
                            public void onComplete(@NonNull Task task) {
                                if (task.isSuccessful()) {
                                    Toast.makeText(ChatActivity.this, "Message Sent Successfully...", Toast.LENGTH_SHORT).show();
                                } else {
                                    Toast.makeText(ChatActivity.this, "Error", Toast.LENGTH_SHORT).show();
                                }
                                loadingBar.dismiss();
                                MessageInputText.setText("");
                            }
                        });
                    }
                });

            } else {
                loadingBar.dismiss();
                Toast.makeText(this, "nothing selected,error", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void DisplayLastSeen() {
        RootRef.child(Constants.NODES.USER_NODE).child(messageReceiverID)
                .addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        if (dataSnapshot.child("userState").hasChild("state")) {
                            String state = dataSnapshot.child("userState").child("state").getValue().toString();
                            String date = dataSnapshot.child("userState").child("date").getValue().toString();
                            String time = dataSnapshot.child("userState").child("time").getValue().toString();

                            if (state.equals("online")) {
                                userLastSeen.setText("online");
                            } else if (state.equals("offline")) {
                                userLastSeen.setText("Last Seen: " + date + " " + time);
                            }
                        } else {
                            userLastSeen.setText("offline");
                        }
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {

                    }
                });
    }


    @Override
    protected void onStart() {
        super.onStart();
        RootRef.child(MESSAGES_NODE).child(Utils.setOneToOneChat(messageSenderID,messageReceiverID))
                .addChildEventListener(new ChildEventListener() {
                    @Override
                    public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                        Messages messages = dataSnapshot.getValue(Messages.class);
                        messagesList.add(messages);
                        messageAdapter.notifyDataSetChanged();
                        userMessagesList.smoothScrollToPosition(userMessagesList.getAdapter().getItemCount());
                    }

                    @Override
                    public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

                    }

                    @Override
                    public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {

                    }

                    @Override
                    public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });
    }

    private void SendMessage() {
        String messageText = MessageInputText.getText().toString();

        if (TextUtils.isEmpty(messageText)) {
            Toast.makeText(this, "first write your message...", Toast.LENGTH_SHORT).show();
        } else {

            DatabaseReference userMessageKeyRef = RootRef.child(MESSAGES_NODE)
                    .child(Utils.setOneToOneChat(messageSenderID,messageReceiverID)).push();
            DatabaseReference userChatNodeKeyRef = RootRef.child(CHAT_NODE);


            String type = "text";
            String messagePushID = userMessageKeyRef.getKey();
            String messageSenderRef = messageSenderID + "/" + messageReceiverID;
            String messageReceiverRef = messageReceiverID + "/" + messageSenderID;
//            String messageRef = "MessagesNode/" + Utils.setOneToOneChat(messageSenderID,messageReceiverID);

            ChatModel chatModel = new ChatModel();
            chatModel.setFromUid(messageSenderID);
            chatModel.setToUid(messageReceiverID);
            chatModel.setDate(saveCurrentDate);
            chatModel.setTime(saveCurrentTime);
            chatModel.setLastMessage(messageText);
            chatModel.setType(type);

            Map messageTextBody = new HashMap();
            messageTextBody.put("message", messageText);
            messageTextBody.put("type", type);
            messageTextBody.put("from", messageSenderID);
            messageTextBody.put("to", messageReceiverID);
            messageTextBody.put("messageID", messagePushID);
            messageTextBody.put("time", saveCurrentTime);
            messageTextBody.put("date", saveCurrentDate);

//            Map messageBodyDetails = new HashMap();
//            messageBodyDetails.put(messageRef + "/" + messagePushID, messageTextBody);

            Map ChatBodyDetails = new HashMap();
            ChatBodyDetails.put(messageSenderRef + "/" , chatModel);
            ChatBodyDetails.put(messageReceiverRef + "/", chatModel);


            userMessageKeyRef.updateChildren(messageTextBody).addOnCompleteListener(new OnCompleteListener() {
                @Override
                public void onComplete(@NonNull Task task) {
                    if (task.isSuccessful()) {
                        Toast.makeText(ChatActivity.this, "Message Sent Successfully...", Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(ChatActivity.this, "Error", Toast.LENGTH_SHORT).show();
                    }
                    MessageInputText.setText("");
                }
            });

            userChatNodeKeyRef.updateChildren(ChatBodyDetails).addOnCompleteListener(new OnCompleteListener() {
                @Override
                public void onComplete(@NonNull Task task) {
                    if (task.isSuccessful()) {
                        Toast.makeText(ChatActivity.this, "Chat Sent Successfully...", Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(ChatActivity.this, "Error", Toast.LENGTH_SHORT).show();
                    }
                }
            });

        }
    }
}

File: app/src/main/java/com/vteam/testdemo/chat/ChatActivity.kt
package com.vteam.testdemo.chat

import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.google.android.gms.tasks.Continuation
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.UploadTask
import com.vteam.testdemo.R
import com.vteam.testdemo.chat.adapter.MessageAdapter
import com.vteam.testdemo.common.Constants
import com.vteam.testdemo.common.Constants.NODES.CHAT_NODE
import com.vteam.testdemo.common.Constants.NODES.MESSAGES_NODE
import com.vteam.testdemo.common.Utils.setOneToOneChat
import com.vteam.testdemo.databinding.ActivityChatBinding
import com.vteam.testdemo.landing.model.ChatModel
import com.vteam.testdemo.top.Messages
import kotlinx.android.synthetic.main.custom_chat_bar.view.*
import java.text.SimpleDateFormat
import java.util.*

class ChatActivity : AppCompatActivity() {
    private lateinit var mBinding: ActivityChatBinding
    private lateinit var uploadTask: UploadTask
    private val messagesList: MutableList<Messages> =
        ArrayList()
    private var messageReceiverID: String? = null
    private var messageReceiverName: String? = null
    private var messageReceiverImage: String? = null
    private var messageSenderID: String? = null

    private var auth: FirebaseAuth? = null
    private var rootRef: DatabaseReference? = null

    private var linearLayoutManager: LinearLayoutManager? = null
    private var messageAdapter: MessageAdapter? = null
//    private var userMessagesList: RecyclerView? = null
    private var loadingBar: ProgressDialog? = null
    private var saveCurrentTime: String? = null
    private var saveCurrentDate: String? = null
    private var checker = ""
    private var myUrl = ""
    private var fileUri: Uri? = null



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)
        mBinding= DataBindingUtil.setContentView(this,R.layout.activity_chat)

        setActionBar()
        auth = FirebaseAuth.getInstance()
        messageSenderID = auth!!.currentUser!!.uid
        rootRef = FirebaseDatabase.getInstance().reference
        val extras = intent.extras
        messageReceiverID = extras!!["visit_user_id"].toString()
        messageReceiverName = extras["visit_user_name"].toString()
        if (extras.containsKey("visit_image") && extras["visit_image"] != null) {
            messageReceiverImage = extras["visit_image"].toString()
        }
        IntializeControllers()
        mBinding.chatToolbar.user_name.text = messageReceiverName
        val reference = FirebaseStorage.getInstance().reference
        if (!TextUtils.isEmpty(messageReceiverImage)) {
            reference.child(messageReceiverImage!!).downloadUrl
                .addOnSuccessListener { uri -> // Got the download URL for 'users/me/profile.png'
                    Log.d("URL", "" + uri)
                    Glide.with(applicationContext)
                        .load(uri)
                        .into(mBinding.chatToolbar.profile_image!!)
                }.addOnFailureListener {
                    // Handle any errors
                }
        }
        mBinding.sendMessageBtn!!.setOnClickListener { SendMessage() }
        DisplayLastSeen()
        mBinding.sendFilesBtn!!.setOnClickListener {
            val options = arrayOf<CharSequence>(
                "Images",
                "Pdf Files",
                "MS Word Files"
            )
            val builder =
                AlertDialog.Builder(this@ChatActivity)
            builder.setTitle("Select the File")
            builder.setItems(options) { dialogInterface, i ->
                if (i == 0) {
                    checker = "image"
                    val intent = Intent()
                    intent.action = Intent.ACTION_GET_CONTENT
                    intent.type = "image/*"
                    startActivityForResult(
                        Intent.createChooser(intent, "Select Image"),
                        443
                    )
                }
                if (i == 1) {
                    checker = "pdf"
                    val intent = Intent()
                    intent.action = Intent.ACTION_GET_CONTENT
                    intent.type = "application/pdf"
                    startActivityForResult(Intent.createChooser(intent, "Select PDF"), 443)
                }
                if (i == 2) {
                    checker = "docx"
                    val intent = Intent()
                    intent.action = Intent.ACTION_GET_CONTENT
                    intent.type = "application/msword"
                    startActivityForResult(
                        Intent.createChooser(
                            intent,
                            "Select MSWORD FILE"
                        ), 443
                    )
                }
            }
            builder.show()
        }
    }

    private fun IntializeControllers() {
        loadingBar = ProgressDialog(this)

        messageAdapter = MessageAdapter(messagesList)

        linearLayoutManager = LinearLayoutManager(this)
        mBinding.userMessagesList!!.layoutManager = linearLayoutManager
        mBinding.userMessagesList!!.adapter = messageAdapter
        val calendar = Calendar.getInstance()
        val currentDate = SimpleDateFormat("MMM dd, yyyy")
        saveCurrentDate = currentDate.format(calendar.time)
        val currentTime = SimpleDateFormat("hh:mm a")
        saveCurrentTime = currentTime.format(calendar.time)
    }

    private fun setActionBar() {

        setSupportActionBar(mBinding.chatToolbar as Toolbar)
        val actionBar = supportActionBar
        actionBar!!.setDisplayShowCustomEnabled(true)
        val layoutInflater =
            this.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val actionBarView =
            layoutInflater.inflate(R.layout.custom_chat_bar, null)
        actionBar.customView = actionBarView
    }

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        data: Intent?
    ) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 443 && resultCode == Activity.RESULT_OK && data != null && data.data != null) {
            loadingBar!!.setTitle("Sending File")
            loadingBar!!.setMessage("Please wait, we are sending....")
            loadingBar!!.setCanceledOnTouchOutside(false)
            loadingBar!!.show()
            fileUri = data.data
            if (checker != "image") {
                val storageReference =
                    FirebaseStorage.getInstance().reference.child("Image Files")
                val messageSenderRef =
                    "Messages/$messageSenderID/$messageReceiverID"
                val messageReceiverRef =
                    "Messages/$messageReceiverID/$messageSenderID"
                val userMessageKeyRef = rootRef!!.child("Messages")
                    .child(messageSenderID!!).child(messageReceiverID!!).push()
                val messagePushID = userMessageKeyRef.key
                val filePath =
                    storageReference.child("$messagePushID.$checker")
                filePath.putFile(fileUri!!).addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        val messageTextBody: MutableMap<String, Any?> =
                            HashMap<String, Any?>()
                        messageTextBody["message"] = myUrl
                        messageTextBody["name"] = fileUri!!.lastPathSegment
                        if (checker == "pdf") {
                            messageTextBody["type"] = checker
                        } else {
                            messageTextBody["type"] = checker
                        }
                        messageTextBody["from"] = messageSenderID
                        messageTextBody["to"] = messageReceiverID
                        messageTextBody["messageID"] = messagePushID
                        messageTextBody["time"] = saveCurrentTime
                        messageTextBody["date"] = saveCurrentDate
                        val messageBodyDetails: MutableMap<String, Any?> =
                            HashMap<String, Any?>()
                        messageBodyDetails["$messageSenderRef/$messagePushID"] = messageTextBody
                        messageBodyDetails["$messageReceiverRef/$messagePushID"] = messageTextBody
                        rootRef!!.updateChildren(messageBodyDetails)
                        loadingBar!!.dismiss()
                    }
                }.addOnFailureListener { e ->
                    loadingBar!!.dismiss()
                    Toast.makeText(this@ChatActivity, e.message, Toast.LENGTH_SHORT).show()
                }
                    .addOnProgressListener { taskSnapshot ->
                        val p =
                            100.0 * taskSnapshot.bytesTransferred / taskSnapshot.totalByteCount
                        loadingBar!!.setMessage((p as Int).toString() + "% Uploading...")
                    }
            } else if (checker == "image") {
                val storageReference =
                    FirebaseStorage.getInstance().reference.child("Image Files")
                val messageSenderRef =
                    "Messages/$messageSenderID/$messageReceiverID"
                val messageReceiverRef =
                    "Messages/$messageReceiverID/$messageSenderID"
                val userMessageKeyRef = rootRef!!.child("Messages")
                    .child(messageSenderID!!).child(messageReceiverID!!).push()
                val messagePushID = userMessageKeyRef.key
                val filePath = storageReference.child("$messagePushID.jpg")

              uploadTask = filePath.putFile(fileUri!!)
                uploadTask.continueWithTask(object :
                    Continuation<UploadTask.TaskSnapshot, Task<Uri>>{
                    override fun then(p0: Task<UploadTask.TaskSnapshot>): Task<Uri> {
                        if (!p0.isSuccessful) {
                            throw p0.exception!!
                        }
                        return filePath.downloadUrl
                    }
                })
                    .addOnCompleteListener { task ->
                        val downloadUrl = task.result
                        myUrl = downloadUrl.toString()
                        val messageSenderRef =
                            "ChatNode/$messageSenderID/$messageReceiverID"
                        val messageReceiverRef =
                            "ChatNode/$messageReceiverID/$messageSenderID"
                        val messageRef =
                            "MessagesNode/" + setOneToOneChat(
                                messageSenderID!!,
                                messageReceiverID!!
                            )
                        val chatModel = ChatModel()
                        chatModel.fromUid = messageSenderID
                        chatModel.fromUid = messageReceiverID
                        chatModel.date = saveCurrentDate
                        chatModel.time = saveCurrentTime
                        chatModel.lastMessage = myUrl
                        chatModel.type = checker
                        val messageTextBody: MutableMap<String, Any?> =
                            HashMap<String, Any?>()
                        messageTextBody["message"] = myUrl
                        messageTextBody["name"] = fileUri!!.lastPathSegment
                        messageTextBody["type"] = checker
                        messageTextBody["from"] = messageSenderID
                        messageTextBody["to"] = messageReceiverID
                        messageTextBody["messageID"] = messagePushID
                        messageTextBody["time"] = saveCurrentTime
                        messageTextBody["date"] = saveCurrentDate
                        val messageBodyDetails: MutableMap<String, Any?> =
                            HashMap<String, Any?>()
                        messageBodyDetails["$messageSenderRef/"] = chatModel
                        messageBodyDetails["$messageReceiverRef/$messagePushID"] = chatModel
                        messageBodyDetails["$messageRef/$messagePushID"] = messageTextBody
                        rootRef!!.updateChildren(messageBodyDetails)
                            .addOnCompleteListener(object : OnCompleteListener<Void> {

                                override fun onComplete(p0: Task<Void>) {
                                    loadingBar!!.dismiss()
                                    mBinding.inputMessage!!.setText("")
                                }
                            })
                    }

            } else {
                loadingBar!!.dismiss()
//                Toast.makeText(this, "nothing selected,error", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun DisplayLastSeen() {
        rootRef!!.child(Constants.NODES.USER_NODE).child(messageReceiverID!!)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.child("userState").hasChild("state")) {
                        val state =
                            dataSnapshot.child("userState").child("state").value.toString()
                        val date =
                            dataSnapshot.child("userState").child("date").value.toString()
                        val time =
                            dataSnapshot.child("userState").child("time").value.toString()
                        if (state == "online") {
                            mBinding.chatToolbar.user_last_seen.text = "online"
                        } else if (state == "offline") {
                            mBinding.chatToolbar.user_last_seen.text = "Last Seen: $date $time"
                        }
                    } else {
                        mBinding.chatToolbar.user_last_seen.text= "offline"
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    override fun onStart() {
        super.onStart()
        rootRef!!.child(MESSAGES_NODE).child(
            setOneToOneChat(
                messageSenderID!!,
                messageReceiverID!!
            )
        )
            .addChildEventListener(object : ChildEventListener {
                override fun onChildAdded(
                    dataSnapshot: DataSnapshot,
                    s: String?
                ) {
                    val messages =
                        dataSnapshot.getValue(
                            Messages::class.java
                        )!!
                    messagesList.add(messages)
                    messageAdapter!!.notifyDataSetChanged()
                    mBinding.userMessagesList!!.smoothScrollToPosition(
                        mBinding.userMessagesList!!.adapter!!.itemCount
                    )
                }

                override fun onChildChanged(
                    dataSnapshot: DataSnapshot,
                    s: String?
                ) {
                }

                override fun onChildRemoved(dataSnapshot: DataSnapshot) {}
                override fun onChildMoved(
                    dataSnapshot: DataSnapshot,
                    s: String?
                ) {
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    private fun SendMessage() {
        val messageText = mBinding.inputMessage.text.toString()
        if (TextUtils.isEmpty(messageText)) {
            Toast.makeText(this, "first write your message...", Toast.LENGTH_SHORT).show()
        } else {
            val userMessageKeyRef = rootRef!!.child(MESSAGES_NODE)
                .child(
                    setOneToOneChat(
                        messageSenderID!!,
                        messageReceiverID!!
                    )
                ).push()
            val userChatNodeKeyRef = rootRef!!.child(CHAT_NODE)
            val type = "text"
            val messagePushID = userMessageKeyRef.key
            val messageSenderRef = "$messageSenderID/$messageReceiverID"
            val messageReceiverRef = "$messageReceiverID/$messageSenderID"
            //            String messageRef = "MessagesNode/" + Utils.setOneToOneChat(messageSenderID,messageReceiverID);
            val chatModel = ChatModel()
            chatModel.fromUid = messageSenderID
            chatModel.toUid = messageReceiverID
            chatModel.date = saveCurrentDate
            chatModel.time = saveCurrentTime
            chatModel.lastMessage = messageText
            chatModel.type = type
            val messageTextBody: MutableMap<String, Any?> =
                HashMap<String, Any?>()
            messageTextBody["message"] = messageText
            messageTextBody["type"] = type
            messageTextBody["from"] = messageSenderID
            messageTextBody["to"] = messageReceiverID
            messageTextBody["messageID"] = messagePushID
            messageTextBody["time"] = saveCurrentTime
            messageTextBody["date"] = saveCurrentDate

//            Map messageBodyDetails = new HashMap();
//            messageBodyDetails.put(messageRef + "/" + messagePushID, messageTextBody);
            val ChatBodyDetails: MutableMap<String, Any?> =
                HashMap<String, Any?>()
            ChatBodyDetails["$messageSenderRef/"] = chatModel
            ChatBodyDetails["$messageReceiverRef/"] = chatModel
            userMessageKeyRef.updateChildren(messageTextBody)
                .addOnCompleteListener(object : OnCompleteListener<Void> {
                    override fun onComplete(task: Task<Void>) {
                       mBinding.inputMessage!!.setText("")
                    }
                })
            userChatNodeKeyRef.updateChildren(ChatBodyDetails)
                .addOnCompleteListener(object : OnCompleteListener<Void> {
                    override fun onComplete(task: Task<Void>) {

                    }
                })
        }
    }
}